tests: Bluetooth: audio: Add initial ASCS unit tests

MariuszSkamra opened the pull request:
This adds initial set of ASCS unit tests and mocks needed to compile the code.
Failing test cases have been conditionally skipped with Zephyr issue number.

to run the tests: `./scripts/twister -v -T tests/bluetooth/audio/ascs --enable-valgrind`

Thalley commented on tests/bluetooth/audio/ascs/mocks/ascs_help_utils.c:
```diff
@@ -0,0 +1,43 @@
+	return NULL;
+}
+
+void ascs_foreach_ase_attr(ascs_foreach_ase_attr_func_t func, void *user_data)
+{
+	const struct bt_gatt_service_static *service = ascs_get_gatt_service();
+	uint8_t result;
+
+	for (size_t i = 0; i < service->attr_count; i++) {
+		const struct bt_gatt_attr *attr = &service->attrs[i];
+
+		if (bt_uuid_cmp(attr->uuid, BT_UUID_ASCS_ASE_SNK) ||
+		    bt_uuid_cmp(attr->uuid, BT_UUID_ASCS_ASE_SRC)) {
+			continue;
+		}
+
+		result = func(attr, user_data);
+		if (result == BT_GATT_ITER_STOP) {
+			return;
+		}
+	}
+}
```
I think this is a duplicate of `bt_audio_unicast_server_foreach_ep`, isn't it?

MariuszSkamra replied:
Not really, as the bt_audio_unicast_server_foreach_ep iterates non-idle endpoints only

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -2653,3 +2647,10 @@ static int control_point_notify(struct bt_conn *conn, const void *data, uint16_t
 
 #endif /* BT_AUDIO_UNICAST_SERVER */
+
+#if defined(ZTEST_UNITTEST)
+const struct bt_gatt_service_static *ascs_get_gatt_service(void)
+{
+	return &ascs_svc;
+}
+#endif /* ZTEST_UNITTEST */
```
Hmm, generally we avoid having test-specific code in the stack, but maybe it's inevitable

Thalley replied:
Hmm, perhaps `bt_gatt_foreach_attr` can be used for this instead?

MariuszSkamra replied:
As this is compiled out of kernel, I decided provide test doubles for GATT. The GATT itself has many kernel dependencies I did not want to mock. I.e here is no `bt_gatt_foreach_attr` function compiled, so I needed to provide a mock for it.

Thalley commented on tests/bluetooth/audio/ascs/mocks/ascs_help_utils.h:
```diff
@@ -0,0 +1,34 @@
+
+#include <zephyr/bluetooth/gatt.h>
+
+#define ASCS_ASE_STATE_IDLE                     0x00
+#define ASCS_ASE_STATE_CODEC_CONFIGURED         0x01
+#define ASCS_ASE_STATE_QOS_CONFIGURED           0x02
+#define ASCS_ASE_STATE_ENABLING                 0x03
+#define ASCS_ASE_STATE_STREAMING                0x04
+#define ASCS_ASE_STATE_DISABLING                0x05
+#define ASCS_ASE_STATE_RELEASING                0x06
```
We have `enum bt_audio_state` in `audio.h`

MariuszSkamra replied:
removed

Thalley reviewed (commented):
Nice approach. Perhaps it would be simpler if we split the changes to the host (and the tests thereof) and the ASCS tests into 2 PRs, as they are not really related, even if there is a dependency. 

The amount of mocking code is generally why we haven't done unit tests for LE Audio yet, but a lot of it can of course be reused. 

I'd like to find a way to avoid introducing test specific code into the stack, so let's see if we can find a way to avoid that

Thalley commented on subsys/bluetooth/audio/audio_internal.h:
```diff
@@ -39,9 +39,14 @@ ssize_t bt_audio_write_chrc(struct bt_conn *conn, const struct bt_gatt_attr *att
 	.user_data = _user_data, \
 }
 
+/* Suppress the function pointer to constant comparison warning */
+#define _IS_NULL(_ptr) ((uintptr_t)(_ptr) == (uintptr_t)NULL)
```
Can you elaborate on this? 

What error/warning do you get by comparing a function pointer in `BT_AUDIO_CHRC` with `NULL` directly? What we simply cast it to `void *`?

MariuszSkamra replied:
I'll probably remove this one

Thalley commented on tests/bluetooth/audio/ascs/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+target_sources(testbinary
+  PRIVATE
+    src/main.c
+    #src/test_suite_gatt_help_utils.c
```
Intentionally commented?

MariuszSkamra replied:
leftover

Thalley commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+	ssize_t ret;
+
+	ret = fixture->ase_snk_1->read(&fixture->conn, fixture->ase_snk_1, NULL, 0, 0);
+	zassert_false(ret < 0, "attr->read returned unexpected (err 0x%02x)",
+		      BT_GATT_ERR(ret));
```
Oh, you want to revert it from a negative value to a `BT_ATT_ERR_` again?

MariuszSkamra replied:
Yes, to print out ATT error

Thalley replied:
Right :) 

Perhaps you should define a new macro for this `GATT_TO_ATT_ERR` that does this, instead of "misusing" the `BT_GATT_ERR` macro

Thalley commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+					     data_expected, ARRAY_SIZE(data_expected));
+}
+
+ZTEST_F(ascs_ase_control_test_suite, test_sink_ase_control_operation_zero_length_write)
+{
+	uint8_t buf[] = {};
+	ssize_t ret;
+
+	ret = fixture->ase_cp->write(&fixture->conn, fixture->ase_cp, (void *)buf, 0, 0, 0);
+	zassert_true(ret < 0, "ase_cp_attr->write returned unexpected (err 0x%02x)",
+		     BT_GATT_ERR(ret));
+}
```
What about a test with `NULL` buf? Or is that expected never to happen?

MariuszSkamra replied:
I would say that this is expected never to happen as per GATT API.

Thalley commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+ZTEST_F(ascs_ase_control_test_suite, test_sink_ase_unsupported_opcode)
+{
+	uint8_t buf[] = {
+		0xFE,   /* Opcode */
```
How is `0xFE` unsuported, but `0x00` invalid? Aren't they both invalid?

MariuszSkamra replied:
To explain myself,
`0x00` is not mentioned, meaning it's invalid
`0x09-0xFF` are RFU
but I agree, this can be put into one test case

Thalley replied:
> 0x00 is not mentioned, meaning it's invalid

Is that true? The spec defines RFU and Prohibited, but I don't see where it states that a undefined value indicates that it is invalid. 

Does the test spec say anything here?

MariuszSkamra replied:
![image](https://user-images.githubusercontent.com/10126877/225664464-4376797e-fdc8-4e3a-aa18-27ae8dbe91fc.png)
![image](https://user-images.githubusercontent.com/10126877/225664535-0e0b082f-b10f-4655-93db-6f7ba541b362.png)
there is no mention about 0x00

Thalley commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+		0x01,           /* Opcode */
+		0xFF,           /* Number_of_ASEs */
+		0x00,           /* ASE_ID[0] */
+		0x02,           /* Response_Code[0] = Invalid Length */
```
Is invalid length what we should expect if number of ASEs = 0?

MariuszSkamra replied:
Yes, the ASE_ID is 0 because the Response_Code is 0x02, as defined in ASCS specification:
 
![image](https://user-images.githubusercontent.com/10126877/225237506-db61e88e-4f4e-47ad-8ffa-d00922336344.png)

Thalley replied:
My comment was not for the ASE_ID, but the Invalid Length response code :)

MariuszSkamra replied:
Not sure what you mean. In request, the ASE_ID is 0x01. In response ASE_ID is 0x00 as per spec

MariuszSkamra replied:
The ASE_ID in `buf` (request) was 0x01. You were probably looking at response, where it's 0x00

fredrikdanebjer commented on tests/bluetooth/audio/ascs/mocks/ascs_help_utils.h:
```diff
@@ -0,0 +1,13 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Is the header guard left out intentionally as this is sort of an internal file? I can see the value in getting the compilation errors if the wrong file includes this, but it also weighs to not see the guard :)

MariuszSkamra replied:
@fredrikdanebjer You know, I end up with guards. It's because I had to include one of the headers in other header file, which caused compilation error for me ;)

MariuszSkamra commented on tests/net/buf_simple/src/main.c:
```diff
@@ -0,0 +1,241 @@
+ZTEST(net_buf_simple_tests, test_net_buf_simple_byte_order)
+{
+	NET_BUF_SIMPLE_DEFINE(buf, 16);
+	uint8_t le16[2] = { 0x02, 0x01 };
```
It would introduce unnecessary dependency on `bswap` in unit tests

MariuszSkamra commented on tests/net/mocks/assert.h:
```diff
@@ -0,0 +1,16 @@
+
+DECLARE_FAKE_VALUE_FUNC(bool, mock_check_if_assert_expected);
+
+#define expect_assert()     (mock_check_if_assert_expected_fake.return_val = 1)
```
See my comment here https://github.com/zephyrproject-rtos/zephyr/pull/55536#discussion_r1130960507

MariuszSkamra commented on tests/net/buf_simple/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+common:
+  tags: test_framework net buf
```
That's a good question actually. Something tells me that `test_framework` is tag that supposed to be used for testing ZTEST test framework (`tests/ztest/`), but it started to be used in other places as well. But as you said, I added this tag, because it is in other bluetooth unit tests. I'll ask on discord

MariuszSkamra commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+{
+	uint8_t data_expected[] = {
+		0x01,   /* ASE_ID */
+		0x00,   /* ASE_State = Idle */
```
I was thinking about it. I would need to define the state values, because I cannot use the same that are defined already in `bap.h`

MariuszSkamra commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+	bap_unicast_server_cb_release_fake.return_val = 0;
+}
+
+static void ascs_ase_control_test_suite_after(void *f)
+{
```
Not necessarily, `ARG_UNUSED` is used to suppress unused variable warning, which is not a case here (`f` is function argument, the compiler won't complain)

MariuszSkamra commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+
+static void ascs_ase_control_test_suite_before(void *f)
+{
+	struct ascs_ase_control_test_suite_fixture *fixture = f;
+
+	ARG_UNUSED(fixture);
```
ah, it's just a leftover

Thalley replied:
@MariuszSkamra please fix :)

MariuszSkamra commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+{
+	Z_TEST_SKIP_IFNDEF(CONFIG_BT_ASCS_ASE_SNK);
+
+	zassert_not_null(ascs_get_attr(BT_UUID_ASCS_ASE_SNK, 1));
+	zassert_not_null(ascs_get_attr(BT_UUID_ASCS_ASE_SNK, CONFIG_BT_ASCS_ASE_SNK_COUNT));
```
There are already places where the same approach has been made, e.g. `subsys/bluetooth/host/buf.c`

MariuszSkamra commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+
+ZTEST_SUITE(ascs_attrs_test_suite, NULL, NULL, NULL, NULL, NULL);
+
+ZTEST(ascs_attrs_test_suite, test_has_sink_ase_chrc)
+{
+	Z_TEST_SKIP_IFNDEF(CONFIG_BT_ASCS_ASE_SNK);
+
+	zassert_not_null(ascs_get_attr(BT_UUID_ASCS_ASE_SNK, 1));
+	zassert_not_null(ascs_get_attr(BT_UUID_ASCS_ASE_SNK, CONFIG_BT_ASCS_ASE_SNK_COUNT));
+}
```
if `CONFIG_BT_ASCS_ASE_SNK_COUNT` is 0 then `CONFIG_BT_ASCS_ASE_SNK` is not set, meaning this test will be skipped, because it's guarded with `Z_TEST_SKIP_IFNDEF(CONFIG_BT_ASCS_ASE_SNK);`

MariuszSkamra commented on tests/bluetooth/audio/ascs/mocks/gatt.c:
```diff
@@ -0,0 +1,118 @@
+/*
```
If there will be a need to reuse them, those can be always moved. YAGNI :D

MariuszSkamra commented on tests/bluetooth/audio/ascs/Kconfig:
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2023 Codecoup
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_CONN
+	bool
+	default y
+
+config BT_MAX_CONN
+	int "Maximum number of simultaneous connections (for unit tests)"
+	depends on BT_CONN
+	range 1 250
+	default 1
+	help
+	  Maximum number of simultaneous Bluetooth connections
+	  supported.
+
+config BT_ISO_MAX_CHAN
+	int "Maximum number of simultaneous ISO channels (for unit test)"
+	default BT_MAX_CONN if BT_CONN
+	default 1
+	range 1 64
+	help
+	  Maximum number of simultaneous Bluetooth isochronous channels
+	  supported.at would make inserting a new
+	  segment into a PDU worthwhile.
```
Hmm, I've seen such approach. Basically it gives an extra possibility to run the tests with different settings defined in `testcase.yaml` e.g.

```
common:
  tags: test_framework bluetooth
tests:
  bluetooth.audio.ascs.iso_max_chan_1:
    type: unit
    extra_configs:
      - CONFIG_BT_ISO_MAX_CHAN=1
  bluetooth.audio.ascs.iso_max_chan_10:
    type: unit
    extra_configs:
      - CONFIG_BT_ISO_MAX_CHAN=10
```

Thalley replied:
> Basically it gives an extra possibility to run the tests with different settings defined in testcase.yaml e.g.

That can still be done (that's how we test the BT shell).

But redefining the Kconfigs here does not only make it harder to maintain, the Kconfig options you define here may also modify the allowed value range, etc. which could in theory allow you to misconfigure the test, perhaps triggering issues that could otherwise not happen.

kruithofa replied:
unittest in tests/bluetooth/controller have been updated recently. 
You should not redefine the Kconfigs, but set them in prj.conf files (see for ex. tests/bluetooth/controller/ctrl_phy_update)

MariuszSkamra replied:
@kruithofa OK

fredrikdanebjer commented:
It might have been mentioned before, but this produced really a huge amount mock files, which in this case is specific to ascs. It would be nice to be able re-use as much as possible for all of le-audio.

Now I have to admit that I haven't looked into detail of all those mocks, but I was wondering if you considered an approach where they are moved up one step and made __weak ? In that way other LE-Audio tests could reuse them, yet easily override them with their specific implementation if needed.

MariuszSkamra commented:
> It might have been mentioned before, but this produced really a huge amount mock files, which in this case is specific to ascs. It would be nice to be able re-use as much as possible for all of le-audio.
> 
> Now I have to admit that I haven't looked into detail of all those mocks, but I was wondering if you considered an approach where they are moved up one step and made __weak ? In that way other LE-Audio tests could reuse them, yet easily override them with their specific implementation if needed.

I don't think I can make those _weak as most of the mocks are created using FFF and macros like:
```
DECLARE_FAKE_VALUE_FUNC(ssize_t, bt_gatt_attr_read, struct bt_conn *,
			const struct bt_gatt_attr *, void *, uint16_t, uint16_t, const void *,
			uint16_t);
```

where one cannot prepend the return type with __weak descriptor, because the macro will evaluate to code that won't compile.

kruithofa commented on tests/bluetooth/audio/ascs/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+  PRIVATE
+    src/main.c
+
+    ${ZEPHYR_BASE}/subsys/bluetooth/audio/ascs.c
```
These files could be part of a library, as was done in controller/uut

MariuszSkamra replied:
done

kruithofa commented on tests/bluetooth/audio/ascs/mocks/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+  ${ZEPHYR_BASE}/tests/bluetooth/audio
+  ${ZEPHYR_BASE}/subsys/bluetooth
+  ${ZEPHYR_BASE}/subsys/bluetooth/audio
+  ${ZEPHYR_BASE}/subsys/bluetooth/audio/ascs/mocks
```
this directory does not exist?

MariuszSkamra replied:
done

kruithofa commented on tests/bluetooth/audio/ascs/mocks/mock-sections.ld:
```diff
@@ -0,0 +1,21 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+SECTIONS
```
why is this necessary?

MariuszSkamra replied:
The unit test linker script seem to be very simple. It does not create sections i.e. for Bluetooth (`common-rom-bt.ld`) so that `bt_gatt_service_static` and `bt_conn_cb` iterable sections are not there.
ASCS uses those 2 sections to register connection callbacks (`bt_conn_cb`) and register service instance (`bt_gatt_service_static`).

Thus I added this script that will create the missing sections

kruithofa commented on tests/bluetooth/audio/ascs/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+  bluetooth.audio.ascs.test:
+    type: unit
+    extra_configs:
+      - CONFIG_BT_ASCS=y
```
move these to the prj.conf file

MariuszSkamra replied:
done

kruithofa commented on tests/bluetooth/audio/ascs/mocks/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+#
```
IMO it would be nice to separate .c and .h files in their own directory src and include
Keep in mind also that sooner or later we should relocate the unittest for host, ascs and controller.

MariuszSkamra replied:
done

MariuszSkamra commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -0,0 +1,767 @@
+	/* The specification rather relaxed in terms of Response_Code value */
+	zassert_true(((uint8_t *)bt_gatt_notify_uuid_arg_data)[3] == 0x07 ||
+		     ((uint8_t *)bt_gatt_notify_uuid_arg_data)[3] == 0x08 ||
+		     ((uint8_t *)bt_gatt_notify_uuid_arg_data)[3] == 0x09, "Invalid Response_Code[0]");
```
removed

MariuszSkamra committed:
Bluetooth: audio: ascs: Add cleanup function

This adds ascs_cleanup() function guarded by ZTEST_UNITTEST. The purpose
of it is to clean up the ASCS between the tests.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: bluetooth: audio: Add mocks for audio

This adds common mocks to be used while unit testing Bluetooth Audio
code.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: audio: Add initial ASCS unit tests

This adds initial ASCS unit tests.
Failing test cases have been conditionally skipped with Zephyr issue
number.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
Changed the name of one of the test cases

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -2674,4 +2674,41 @@ static int control_point_notify(struct bt_conn *conn, const void *data, uint16_t
 }
 
+#if defined(ZTEST_UNITTEST)
+static void ase_cleanup(struct bt_ascs_ase *ase)
+{
+	if (ase->ep.iso != NULL) {
+		bt_bap_iso_unbind_ep(ase->ep.iso, &ase->ep);
+	}
+
+	if (ase->ep.stream != NULL) {
+		bt_bap_stream_detach(ase->ep.stream);
+	}
+
+	bt_ascs_ase_return_to_slab(ase);
+}
+
+void ascs_cleanup(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(sessions); i++) {
+		struct bt_ascs *session = &sessions[i];
+		sys_snode_t *node, *tmp;
+
+		if (session->conn == NULL) {
+			continue;
+		}
+
+		SYS_SLIST_FOR_EACH_NODE_SAFE(&session->ases, node, tmp) {
+			struct bt_ascs_ase *ase;
+
+			ase = CONTAINER_OF(node, struct bt_ascs_ase, node);
+
+			ase_cleanup(ase);
+		}
+
+		bt_conn_unref(session->conn);
+		session->conn = NULL;
+	}
+}
```
Can you elaborate on why these are needed? 
Could the same behavior be done if we could e.g. trigger a disconnection callback?

MariuszSkamra replied:
Yes, it probably can, however it's implementation detail. The disconnected callback actually does more than the cleanup function needs to do (transition to idle, sending notifications, calling application callbacks and so on). Moreover, the implementation may change, so that disconnect callback won't do the cleanup anymore (e.g. some caching will be done).
Using dedicated cleanup function makes the tests not bound with implementation.

Thalley commented on tests/bluetooth/audio/mocks/include/gatt.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef MOCKS_GATT_H_
+#define MOCKS_GATT_H_
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/uuid.h>
```
Wouldn't it be enough to include `gatt.h` here?

Thalley commented on tests/bluetooth/audio/mocks/include/gatt_expects.h:
```diff
@@ -0,0 +1,114 @@
+#ifndef MOCKS_GATT_EXPECTS_H_
+#define MOCKS_GATT_EXPECTS_H_
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/uuid.h>
```
Wouldn't it be enough to include `gatt.h` here?

Thalley commented on tests/bluetooth/audio/mocks/include/iso.h:
```diff
@@ -0,0 +1,26 @@
+#define MOCKS_ISO_H_
+
+#include <zephyr/fff.h>
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/iso.h>
+#include <zephyr/net/buf.h>
```
Ditto with just `iso.h` here? 

How come this includes `fff.h`, but the GATT files do not?

MariuszSkamra replied:
fff.h in gatt.h was missing

Thalley commented on tests/bluetooth/audio/mocks/include/mock_kernel.h:
```diff
@@ -0,0 +1,32 @@
+#ifndef MOCKS_KERNEL_H_
+#define MOCKS_KERNEL_H_
+
+#include <zephyr/fff.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys_clock.h>
```
What is included from `sys_clock.h` here?

MariuszSkamra replied:
k_ticks_t

Thalley commented on tests/bluetooth/audio/mocks/include/pacs.h:
```diff
@@ -0,0 +1,21 @@
+#define MOCKS_PACS_H_
+
+#include <zephyr/fff.h>
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/audio/audio.h>
+#include <zephyr/bluetooth/audio/pacs.h>
```
Ditto with just `pacs.h` here, right?

Thalley commented on tests/bluetooth/audio/mocks/src/bap_stream.c:
```diff
@@ -0,0 +1,59 @@
+	mock_bap_stream_ops.sent = mock_bap_stream_sent_cb;
+}
+
+void mock_bap_stream_cleanup(void)
+{
+
+}
```
What is the purpose of this and the other empty functions?

MariuszSkamra replied:
Simple, but defined mock interface *_init, *_cleanup

Thalley replied:
That's also what I thought :)

Thalley commented on tests/bluetooth/audio/mocks/src/gatt.c:
```diff
@@ -0,0 +1,223 @@
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	*num_matches -= 1;
```
Any chance this is 0 when this function is called?

MariuszSkamra replied:
No. Check the line 194

Thalley replied:
Right, so first time this is called, it will always be non-0, but we are decrementing each time it is called, and we only ever check the value again. 

So if the value is `5` the first time, and `gatt_foreach_iter` is called e.g. 6 times where the above check passes, don't we risk this underflowing?

Thalley replied:
At least to me it seems like we need to add a check for `*num_matches == 0` in this.

MariuszSkamra replied:
There is already such check few lines below (176)

Thalley replied:
No, the function that calls this cannot be called with `num_matches == 0`, but `num_matches` is decremented multiple times, and there doesn't seem to be anything stopping it from decrementing it from 3 -> 2 -> 1 -> 0 -> Underflow

MariuszSkamra replied:
Let's consider num_matches  = 1;
```
        *num_matches -= 1;

        // now *num_matches is 0;

	result = func(attr, handle, user_data);

	if (!*num_matches) {
		return BT_GATT_ITER_STOP;    // <- return
	}
```

Thalley replied:
Ah, I finally see it (might be the indention that causes this to be hard to read), as I had not seen that in case of 

```c
	if (!*num_matches) {
		return BT_GATT_ITER_STOP;
	}
```

that would trigger a `return` in `bt_gatt_foreach_attr_type`. I had read that `return` as a `break`, which would not work, but a `return` does. 

Nevermind my comment then :) (but perhaps we should split this function into a few smaller ones in the host (and here) to reduce the indentation level to increase readability.

Thalley commented on tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:
```diff
@@ -0,0 +1,116 @@
+				       "'%s()' was called with incorrect '%s' value",              \
+				       func_name, "ep");))                                         \
+												   \
+	IF_NOT_EMPTY(_dir, (                                                                       \
+		     zassert_equal(_dir, mock_bap_unicast_server_cb_config_fake.arg2_val,          \
+				       "'%s()' was called with incorrect '%s' value",              \
+				       func_name, "_dir");))                                       \
```
I'm having a little trouble understanding the `IF_NOT_EMPTY` macro. 

What does "empty" mean in this context, and how can a value like `_dir` be empty?

MariuszSkamra replied:
`EMPTY` is the macro that has no replacement expression, like:
```
#define EMPTY
#define ANY
```
the `EMPTY` is defined in `include/zephyr/sys/util_macro.h`.

The `IF_NOT_EMPTY` definition evaluates to the code below:
```
#define IF_NOT_EMPTY(_a, _code) COND_CODE_1(UTIL_BOOL(IS_EMPTY(_a)), (), _code)
```
and works like this: `IS_EMPTY(_a)` evaluates to `true` or `false` depending on `_a` has replacement expression. `_a` can be variable, macro, etc. `UTIL_BOOL(_x)` evaluates to 1 or 0. Finally, `COND_CODE_1` makes the `_code` to be compiled only if `_a` has replacement expression.

The purpose of it is to have a possibility to verify only a subset of function parameters under test.

Consider, the function `foo(a, b)` is expected to be called. Parameter `a` is the one we want to verify, but we want to skip `b` verification. With the use of `IF_NOT_EMPTY` one can write a verification function that will skip the parameter verification
```
#define SKIP_THIS_ONE
....
    expect_foo_too_be_called(a, SKIP_THIS_ONE);
```

Thalley replied:
Ah, thanks! That makes sense to me now :)

Thalley commented on tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:
```diff
@@ -0,0 +1,116 @@
+				       func_name, "_dir");))                                       \
+												   \
+	IF_NOT_EMPTY(_codec, (                                                                     \
+		     /* TODO */                                                                    \
```
Is this TODO still valid?

Thalley commented on tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:
```diff
@@ -0,0 +1,116 @@
+				       func_name, "stream");))                                     \
+												   \
+	IF_NOT_EMPTY(_qos, (                                                                       \
+		     /* TODO */                                                                    \
```
Is this TODO still valid?

MariuszSkamra replied:
ditto

Thalley commented on tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:
```diff
@@ -0,0 +1,116 @@
+				       func_name, "stream");))                                     \
+												   \
+	IF_NOT_EMPTY(_meta, (                                                                      \
+		     /* TODO */                                                                    \
```
Is this TODO still valid?

MariuszSkamra replied:
ditto

Thalley commented on tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:
```diff
@@ -0,0 +1,116 @@
+		     zassert_unreachable("Not implemented");))                                     \
+												   \
+	IF_NOT_EMPTY(_meta_count, (                                                                \
+		     /* TODO */                                                                    \
```
Is this TODO still valid?

MariuszSkamra replied:
ditto

Thalley commented on tests/bluetooth/audio/mocks/include/bap_unicast_server_expects.h:
```diff
@@ -0,0 +1,116 @@
+do {                                                                                               \
+	const char *func_name = "bt_bap_unicast_server_cb.release";                                \
+												   \
+	zassert_true(mock_bap_unicast_server_cb_release_fake.call_count > 0,                       \
+		     "'%s()' was not called", func_name);                                          \
+	zassert_equal(1, mock_bap_unicast_server_cb_release_fake.call_count,                       \
+		      "'%s()' was called more than once", func_name);                              \
```
You have this pattern many places. Could it be replaced by 

```c
	zassert_equal(1, mock_bap_unicast_server_cb_release_fake.call_count,                       \
		      "'%s()' was not called just once: %u", func_name, mock_bap_unicast_server_cb_release_fake.call_count); 
```

If we just want to check against the value `1`, it seems superfluous to first check if it is above 0 before checking if it equal to 1.

Thalley replied:
@MariuszSkamra Are you planning on changing this? If you are, then I'll approve once it has been updated. Since this is just a minor thing, I can also improve as is, if you don't want to change this

MariuszSkamra replied:
As you don't insist, let's leave it as it is

MariuszSkamra replied:
Ah I understand now the game you play. There you go, fixed this

Thalley commented on tests/bluetooth/audio/mocks/include/expects_util.h:
```diff
@@ -0,0 +1,32 @@
+static inline void expect_data_equal(const char *func_name, const char *arg_name,
+				     const uint8_t *expect, const uint8_t *data, size_t len)
+{
+	for (size_t i = 0U; i < len; i++) {
+		zexpect_equal(expect[i], data[i],
```
Did you consider `zexpect_mem_equal`?

MariuszSkamra replied:
Yes, I used it, but in this particular case iteration gives more information about the PDU entries that are different than expected. It's easier to catch the problem looking into the logs

Thalley replied:
That makes sense. That could be added as a comment in the code :)

Have you considered casting the data to a struct and have a struct compare instead? That may be even better for debugging purposes

MariuszSkamra replied:
I think it's sufficient as it is. The expected and test data are just uint8_t buffers so it's quite easy to figure out the data that failed based on array index.

Thalley commented on tests/bluetooth/audio/mocks/include/gatt_expects.h:
```diff
@@ -0,0 +1,111 @@
+#include "gatt.h"
+#include "expects_util.h"
+
+#define expect_bt_gatt_attr_read_called_once(_conn, _attr, _buf, _buf_len, _offset, _value,        \
+					     _value_len)                                           \
+do {                                                                                               \
```
Did you consider implementing these as functions instead of these large macros?

MariuszSkamra replied:
I could not use the `IF_NOT_EMPTY` macro then

Thalley commented on tests/bluetooth/audio/mocks/include/gatt_expects.h:
```diff
@@ -0,0 +1,111 @@
+				       "'%s()' was called with incorrect '%s' value",              \
+				       func_name, "conn");))                                       \
+												   \
+	params = mock_bt_gatt_notify_cb_fake.arg1_val;                                             \
```
Do we need a IF_NOT_EMPTY check for `arg1_val` before assigning it to `params` and using `params` right after?

MariuszSkamra replied:
No, as it's ensured to be valid pointer. `params` are stack allocated, thus the `bt_gatt_notify_cb` mock function does deep copy of it and ensures the `bt_gatt_notify_cb` was called with valid parameters:
audio/mocks/src/gatt.c
```
	zassert_not_null(params, "'%s()' was called with incorrect '%s' value", __func__, "params");

	/* Either params->uuid, params->attr, or both has to be provided */
	zassert_true(params->uuid != NULL || params->attr != NULL,
		     "'%s()' was called with incorrect '%s' value", __func__,
		     "params->uuid or params->attr");
```

Thalley commented on tests/bluetooth/audio/mocks/src/fatal.c:
```diff
@@ -0,0 +1,13 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/fatal.h>
+#include <zephyr/kernel.h>
+
+void z_fatal_error(unsigned int reason, const z_arch_esf_t *esf)
+{
+	ztest_test_fail();
+}
```
Is `kernel.h` needed?
Should we include `ztest.h` since we are using `ztest_test_fail`?

`ztest.h` may be missing from several files

MariuszSkamra replied:
A little trick has been made. In CMakeLists.txt there is a line added:
```
target_compile_options(uut PRIVATE -std=c11 -include ztest.h)
```

`-include` is GCC preprocessor option to include the given header https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html :) I copied it from other place (/tests/bluetooth/host/CMakeLists.txt)

Thalley replied:
If this is a common approach in Zephyr, then that's fine :D It does somewhat go against the concept of "include what you use", but it's fine :)

MariuszSkamra commented:
As https://github.com/zephyrproject-rtos/zephyr/pull/55926 and https://github.com/zephyrproject-rtos/zephyr/pull/55995 were merged I enabled the related test cases

MariuszSkamra commented on tests/bluetooth/audio/mocks/src/conn.c:
```diff
@@ -0,0 +1,30 @@
+	return 0;
+}
+
+struct bt_conn *bt_conn_ref(struct bt_conn *conn)
+{
+	return conn;
+}
+
+void bt_conn_unref(struct bt_conn *conn)
+{
+
+}
```
If there will be a need, those could be easily changed to FFF fake functions.

MariuszSkamra commented on tests/bluetooth/audio/mocks/src/iso.c:
```diff
@@ -0,0 +1,14 @@
+
+#include "iso.h"
+
+DEFINE_FAKE_VALUE_FUNC(int, bt_iso_chan_send, struct bt_iso_chan *, struct net_buf *, uint16_t,
```
I am not sure. the code compiles, the `fff.h` is included in `iso.h` header

Thalley replied:
Yeah, it just a matter of whether we want to be explicit with our includes or not :)

Thalley reviewed (approved):
All blocking comments have been resolved
